Revamp Internal Horizon Documentation (#4948)
stellar/go: services/horizon/internal/docs/GUIDE_FOR_DEVELOPERS.md

```

The return value from `WithField` or `WithFields` is a `*log.Entry`, which you can save to emit multiple logging
statements that all share the same field.  For example, the action system for Horizon attaches a log entry to `action.Log` on every request that can be used to emit log entries that have the request's id attached as a field.

### **Logging and Context**

The logging package in Go provides a default logger, but it may be necessary to include request-specific fields in log statements. This can be achieved by using a context parameter instead of passing a *http.Request to each subroutine. The context can be bound to a logger with log.Set and retrieved with log.Ctx(ctx) to enable logging on behalf of a server request. This approach allows for easier debugging by filtering log streams based on request IDs. The [Go blog](https://blog.golang.org/context) provides more information on using contexts.

Here's an example of using context:

```go

// create a new sublogger
sub := log.WithField("val", 1)

// bind it to a context
ctx := log.Set(context.Background(), sub)

log.Info("no fields on this statement")
log.Ctx(ctx).Info("This statement will use the sub logger")

```

> 

Reviewer: ```suggestion
### **Code Format**

some basic code formatting is required for contributions, scripts are included in this repo to perform these tasks, some of this formatting can be done automatically by IDE's also, these scripts are good way to verify in any case:

./gofmt.sh
./govet.sh
./staticcheck.sh

```

# Horizon Development Guide

This document describes how to build Horizon from source, so that you can test and edit the code locally to develop bug fixes and new features.

## Dependencies
- A [Unix-like](https://en.wikipedia.org/wiki/Unix-like) operating system with the common core commands (cp, tar, mkdir, bash, etc.)
- Go (this repository is officially supported on the last two releases of Go)
- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) (to check out Horizon's source code)
- [mercurial](https://www.mercurial-scm.org/) (needed for `go-dep`)
- [Docker](https://www.docker.com/products/docker-desktop)

## The Go Monorepo
All the code for Horizon resides in our Go monorepo. Fork the repository and clone the fork into your local machine.
```bash
git clone https://github.com/<your-github-username>/go.git
```

> ## The Start script

Reviewer: ```suggestion
## Running Horizon from source
```

The script takes one optional parameter which configures the Stellar network used by the docker containers. If no parameter is supplied, the containers will run on the Stellar test network.

`./start.sh pubnet` will run the containers on the Stellar public network.

`./start.sh standalone` will run the containers on a private standalone Stellar network.

`./start.sh testnet` will run the containers on the Stellar test network.

The following ports will be exposed:
- Horizon: **8000**
- Horizon-Postgres: **5432**
- Stellar-Core (If `standalone` specified): **11626**
- Stellar-Core-Postgres: **5641**

Note that when you switch between different networks you will need to clear the Stellar Core and Stellar Horizon databases. You can wipe out the databases by running `docker-compose down --remove-orphans -v`.

This script is helpful to spin up the services quickly and play around with them. However, for code development it's important to build and install everything locally 

> ## Local Horizon Installation

Reviewer: ```suggestion
## Developing Horizon locally
```

### Building Horizon

1. Change to the horizon services directory - `cd go/services/horizon/`.
2. Compile the Horizon binary: `go build -o stellar-horizon && go install`. You should see the resulting `stellar-horizon` executable in `go/services/horizon`.
3. Add the executable to your PATH in your `~/.bashrc` or equivalent, for easy access: `export PATH=$PATH:{absolute-path-to-horizon-folder}`

Open a new terminal. Confirm everything worked by running `stellar-horizon --help` successfully. You should see an informative message listing the command line options supported by Horizon.

### Database Setup

Horizon uses a Postgres database backend to record information ingested from an associated Stellar Core. The unit and integration tests will also attempt to reference a Postgres db server at ``localhost:5432`` with trust auth method enabled by default for ``postgres`` user.  You can either install the server locally or run any type of docker container that hosts the database server. Here is one example:
```bash

> docker run --platform linux/amd64 -d --env POSTGRES_HOST_AUTH_METHOD=trust -p 5432:5432 circleci/postgres:12-alpine

Reviewer: for consideration, we can be more opinionated towards one suggestion of image to use for db, to reduce content, and just suggest  using `docker compose -f ./docker/docker-compose.yml up horizon-postgres`

Author: @sreuland Weird, since I remember removing that other command and use just this one. Might have reverted that change by mistake

```bash
cd /go/services/horizon/
go test ./...
```

To run the integration tests, you need to set some environment variables:

```bash
export HORIZON_INTEGRATION_TESTS_ENABLED=true 
export HORIZON_INTEGRATION_TESTS_CORE_MAX_SUPPORTED_PROTOCOL=19
export HORIZON_INTEGRATION_TESTS_DOCKER_IMG=stellar/stellar-core:latest
go test -race -timeout 25m -v ./services/horizon/internal/integration/...
```
Note that this will also require a Postgres instance running on port 5432 either locally or exposed through a docker container. Also note that the ``POSTGRES_HOST_AUTH_METHOD`` has been enabled.

### Setup Debug Configuration in IDE

#### Code Debug
Add a debug configuration in your IDE to attach a debugger to the local Horizon process and set breakpoints in your code. Here is an example configuration for VS Code:

```json
 {
     "name": "Horizon Debugger",
     "type": "go",
     "request": "launch",
     "mode": "debug",
     "program": "${workspaceRoot}/services/horizon/main.go",
     "env": {
         "DATABASE_URL": "postgres://postgres@localhost:5432/horizon?sslmode=disable",
         "NETWORK":  "testnet",
         "PER_HOUR_RATE_LIMIT": "0"
     },
     "args": []
 }
```

> If all is well, you should see ingest logs written to standard out.

Reviewer: maybe add a link to the public html docs for references on usage of env settings, such as

https://developers.stellar.org/docs/run-api-server/configuring

noticed this is referencing the new `NETWORK` which is cool, but it's not available on the public docs yet, probably best to refer to the existing settings instead for captive core, passphrase, etc that are mentioned on current docs page. and this can be updated later for the new `NETWORK` when in code and pub'd on docs.

> #### Test Debug

Reviewer: nice, very helpful to include this type of reference for interactive debugging from ide.

```json
{
   "name": "Debug Test Function",
   "type": "go",
   "request": "launch",
   "mode": "test",
   "program": "${workspaceRoot}/services/horizon/internal/integration",
   "env": {
       "HORIZON_INTEGRATION_TESTS_ENABLED": "true",
       "HORIZON_INTEGRATION_TESTS_CORE_MAX_SUPPORTED_PROTOCOL": "19",
       "HORIZON_INTEGRATION_TESTS_DOCKER_IMG": "stellar/stellar-core:latest"
   },
   "args": [
       "-test.run",
       "TestFilteringAccountWhiteList",
       "-test.timeout",
       "5m",
       "./..."
   ],
   "showLog": true
}
```

## Testing Horizon API using Stellar Laboratory

You can test your Horizon instance with a query like: `http://localhost:8000/transactions?limit=10&order=asc`. However, it's much easier to use the [Stellar Laboratory](https://www.stellar.org/laboratory/) to craft other queries to try out. Select `Custom` Horizon URL and enter `http://localhost:8000`.

Read about the available endpoints and see examples in the [Horizon API reference](https://www.stellar.org/developers/horizon/reference/).

# **Notes for Developers**

This section contains additional information related to the development of Horizon.

## Configuring a Standalone Stellar-Core

> By default, the Docker Compose file configures Stellar Core to connect to the Stellar test network.

Reviewer: ```suggestion
By default, the (docker-compose.yml)[(/services/horizon/docker/docker-compose.yml)] will configure Horizon with captive core ingestion to use (Stellar Testnet)[https://developers.stellar.org/docs/fundamentals-and-concepts/testnet-and-pubnet#testnet] network.
```

To run the containers on a private stand-alone network, run `./start.sh standalone`.

> When you run Stellar Core on a private stand-alone network, an account will be created which will hold 100 billion Lumens.

Reviewer: ```suggestion
When you run Stellar Core on a stand-alone network with passphrase of `Standalone Network ; February 2017`, the root account will be created by default with a balance of 100 billion Lumens and a key pair of:
Root Public Key: GBZXN7PIRZGNMHGA7MUUUF4GWPY5AYPV6LY4UV2GL6VJGIQRXFDNMADI
Root Secret Key: SC5O7VZUXDJ6JBDSZ74DSERXL7W3Y5LTOAMRF7RQRL3TAGAPS7LUVG3L

```

can delete lines 145-149

# Horizon Development Guide

This document describes how to build Horizon from source, so that you can test and edit the code locally to develop bug fixes and new features.

## Dependencies
- A [Unix-like](https://en.wikipedia.org/wiki/Unix-like) operating system with the common core commands (cp, tar, mkdir, bash, etc.)

> - Go (this repository is officially supported on the last two releases of Go)

Reviewer 2: Can we specify the supported Go version numbers?

- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) (to check out Horizon's source code)
- [mercurial](https://www.mercurial-scm.org/) (needed for `go-dep`)
- [Docker](https://www.docker.com/products/docker-desktop)

## The Go Monorepo

> All the code for Horizon resides in our Go monorepo. Fork the repository and clone the fork into your local machine.

Reviewer: don't need to fork the repo yet at this point, could this just simply state to do:
```
git clone https://github.com/stellar/go.git
```

so, they can get rolling, and mention if intention is to contribute, then consider forking the repo first and checking that out?

Author: Ok. Yeah I had contribution in mind so included forking it.

```bash
git clone https://github.com/<your-github-username>/go.git
```

## Getting Started with Running Horizon
The [start.sh](/services/horizon/docker/start.sh) script builds horizon from current source, and then runs docker-compose to start the docker containers with runtime configs for horizon, postgres, and optionally core if the optional `standalone` network parameter was included. 
The script takes one optional parameter which configures the Stellar network used by the docker containers. If no parameter is supplied, the containers will run on the Stellar test network.

`./start.sh pubnet` will run the containers on the Stellar public network.

`./start.sh standalone` will run the containers on a private standalone Stellar network.

`./start.sh testnet` will run the containers on the Stellar test network.

The following ports will be exposed:
- Horizon: **8000**
- Horizon-Postgres: **5432**
- Stellar-Core (If `standalone` specified): **11626**

> - Stellar-Core-Postgres: **5641**

Reviewer 2: does `Stellar-Core-Postgres` container run when using testnet or pubnet?

Author: Actually only when standalone is specified. I will add that comment there

Note that when you switch between different networks you will need to clear the Stellar Core and Stellar Horizon databases. You can wipe out the databases by running `docker-compose down --remove-orphans -v`. 

This script is helpful to spin up the services quickly and play around with them. However, for code development it's important to build and install everything locally 

## Developing Horizon Locally
We will now configure a development environment to run Horizon service locally without Docker.

### Building Stellar Core

> Horizon requires a local instance of stellar-core called Captive Core. Since, we are doing this for dev purposes it's a good idea to build it from scratch using the latest release of stellar-core. Head over to the [INSTALL.md](https://github.com/stellar/stellar-core/blob/master/INSTALL.md) file for the instructions.

Reviewer: ```suggestion
Horizon requires an instance of stellar-core binary on the same host. This is referred to as the `Captive Core`. Since, we are running horizon for dev purposes, we recommend considering two approaches to get the stellar-core binary, if saving time is top priority and your development machine is on a linux debian o/s, then consider installing the debian package, otherwise the next option available is to compile the core source directly to binary on your machine, refer to [INSTALL.md](https://github.com/stellar/stellar-core/blob/master/INSTALL.md) file for the instructions on both approaches.
```

### Building Horizon

1. Change to the horizon services directory - `cd go/services/horizon/`.
2. Compile the Horizon binary: `go build -o stellar-horizon && go install`. You should see the resulting `stellar-horizon` executable in `go/services/horizon`.

> 3. Add the executable to your PATH in your `~/.bashrc` or equivalent, for easy access: `export PATH=$PATH:{absolute-path-to-horizon-folder}`

Reviewer: can you just defer to telling user this assumes user has already setup GOPATH in their PATH as part of GO install, and `go install. ` should put the binary in there, so, it should resolve without need to change PATH any further.

Open a new terminal. Confirm everything worked by running `stellar-horizon --help` successfully. You should see an informative message listing the command line options supported by Horizon.

### Database Setup

> Horizon uses a Postgres database backend to record information ingested from an associated Stellar Core. The unit and integration tests will also attempt to reference a Postgres db server at ``localhost:5432`` with trust auth method enabled by default for ``postgres`` user.  You can either install the server locally or run any type of docker container that hosts the database server. We recommend using the [docker-compose.yml](/services/horizon/docker/docker-compose.yml) file in the ``docker`` folder:

Reviewer 2: ```suggestion
Horizon uses a PostgreSQL database backend to store information ingested from the associated Stellar Core. Both unit and integration tests rely on a PostgreSQL database server running at "localhost:5432". You have two options to set up the PostgreSQL server:

1. Use the provided docker container by navigating to "go/services/horizon/docker/" and running:
```bash
docker-compose -f docker-compose.yml up -d horizon-postgres
``

2. Alternatively, you can install a local copy of PostgreSQL and follow the [documentation](https://developers.stellar.org/docs/run-api-server/configuring#preparing-the-database) to prepare the database for use with Horizon.

```

Reviewer 2: Remove lines 55:58

> Horizon uses a Postgres database backend to record information ingested from an associated Stellar Core. The unit and integration tests will also attempt to reference a Postgres db server at ``localhost:5432`` with trust auth method enabled by default for ``postgres`` user.  You can either install the server locally or run any type of docker container that hosts the database server. We recommend using the [docker-compose.yml](/services/horizon/docker/docker-compose.yml) file in the ``docker`` folder:

Reviewer 2: What is the significance of having trust auth method enabled?

Author: From the postgres documentation - "When trust authentication is specified, PostgreSQL assumes that anyone who can connect to the server is authorized to access the database with whatever database user name they specify (even superuser names). Of course, restrictions made in the database and user columns still apply"

Author: I think this avoids having to do the extra steps of adding superuser privileges to another username.

> Horizon uses a Postgres database backend to record information ingested from an associated Stellar Core. The unit and integration tests will also attempt to reference a Postgres db server at ``localhost:5432`` with trust auth method enabled by default for ``postgres`` user.  You can either install the server locally or run any type of docker container that hosts the database server. We recommend using the [docker-compose.yml](/services/horizon/docker/docker-compose.yml) file in the ``docker`` folder:

Reviewer 2: Do we want to include running PostgreSQL using docker here given that the heading explicitly states, "configure a development environment to run Horizon locally without Docker." ?

>You can either install the server locally or run any type of docker container that hosts the database server.

Is there any additional prep required to set up the db in this case?

Author: @urvisavla I meant that we run Horizon individually without Docker but the postgres will always be running on Docker which I have specified how to

```bash

> docker compose -f ./docker/docker-compose.yml up horizon-postgres

Reviewer 2: ```suggestion
docker-compose -f ./docker/docker-compose.yml up horizon-postgres
```

```
This starts a Horizon Postgres docker container and exposes it on the port 5432.

### Run tests
At this point you should be able to run Horizon's unit tests:
```bash

> cd /go/services/horizon/

Reviewer 2: ```suggestion
cd go/services/horizon/
```

```json
 {
     "name": "Horizon Debugger",
     "type": "go",
     "request": "launch",
     "mode": "debug",
     "program": "${workspaceRoot}/services/horizon/main.go",
     "env": {
         "DATABASE_URL": "postgres://postgres@localhost:5432/horizon?sslmode=disable",
         "CAPTIVE_CORE_CONFIG_APPEND_PATH": "./services/horizon/internal/configs/captive-core-testnet.cfg",
         "HISTORY_ARCHIVE_URLS": "https://history.stellar.org/prd/core-testnet/core_testnet_001,https://history.stellar.org/prd/core-testnet/core_testnet_002",
         "NETWORK_PASSPHRASE": "Test SDF Network ; September 2015",
         "PER_HOUR_RATE_LIMIT": "0"
     },
     "args": []
 }
```
If all is well, you should see ingest logs written to standard out. You can read more about configuring the different environment variables in [Configuring](https://developers.stellar.org/docs/run-api-server/configuring) section of our public documentation.

#### Test Debug
You can also use a similar configuration to debug the integration and unit tests. For e.g. here is a configuration for debugging the ```TestFilteringAccountWhiteList``` integration test.

> ```json

Reviewer: this is great, have you tried this snippet i VSCode IDE just to confirm it works?

Author: @sreuland Yes I have confirmed this myself. It works fine in VS Code

   "type": "go",
   "request": "launch",
   "mode": "test",
   "program": "${workspaceRoot}/services/horizon/internal/integration",
   "env": {
       "HORIZON_INTEGRATION_TESTS_ENABLED": "true",
       "HORIZON_INTEGRATION_TESTS_CORE_MAX_SUPPORTED_PROTOCOL": "19",
       "HORIZON_INTEGRATION_TESTS_DOCKER_IMG": "stellar/stellar-core:19.11.0-1323.7fb6d5e88.focal"
   },
   "args": [
       "-test.run",
       "TestFilteringAccountWhiteList",
       "-test.timeout",
       "5m",
       "./..."
   ],
   "showLog": true
}
```

## Testing Horizon API using Stellar Laboratory

You can test your Horizon instance with a query like: `http://localhost:8000/transactions?limit=10&order=asc`. However, it's much easier to use the [Stellar Laboratory](https://www.stellar.org/laboratory/) to craft other queries to try out. Select `Custom` Horizon URL and enter `http://localhost:8000`.

Read about the available endpoints and see examples in the [Horizon API reference](https://www.stellar.org/developers/horizon/reference/).

# **Notes for Developers**

This section contains additional information related to the development of Horizon.

## Configuring a Standalone Stellar-Core

> By default, the [docker-compose.yml](/services/horizon/docker/docker-compose.yml) will configure Horizon with captive core ingestion to use [Stellar Testnet](https://developers.stellar.org/docs/fundamentals-and-concepts/testnet-and-pubnet#testnet) network.

Reviewer 2: I think it might be more helpful to add this link https://developers.stellar.org/docs/fundamentals-and-concepts/testnet-and-pubnet#testnet in # "Getting Started with Running Horizon" section and remove this line from here.

To run the containers on a private stand-alone network, run `./start.sh standalone`.

> When you run Stellar Core on a stand-alone network with passphrase of `Standalone Network ; February 2017`, the root account will be created by default with a balance of 100 billion Lumens and a key pair of:

Reviewer 2: ```suggestion
When you run Stellar Core on a stand-alone network, a root account will be created by default. It will have a balance of 100 billion Lumens and the following key pair:
```

Reviewer 2: The mention of the passphrase here might be confusing since it gives the impression that users need to explicitly provide the passphrase, whereas it is already present in the `stellar-core-standalone.cfg` file.

```
Root Public Key: GBZXN7PIRZGNMHGA7MUUUF4GWPY5AYPV6LY4UV2GL6VJGIQRXFDNMADI
Root Secret Key: SC5O7VZUXDJ6JBDSZ74DSERXL7W3Y5LTOAMRF7RQRL3TAGAPS7LUVG3L
```
When running Horizon on a private stand-alone network, Horizon will not start ingesting until Stellar Core creates its first history archive snapshot. Stellar Core creates snapshots every 64 ledgers, which means ingestion will be delayed until ledger 64.

### Accelerated network for testing

You can increase the speed of tests by adding `ARTIFICIALLY_ACCELERATE_TIME_FOR_TESTING=true` in the [captive-core-standalone.cfg](/services/horizon/docker/captive-core-standalone.cfg).

And finally setting the checkpoint frequency for horizon:
```bash
export CHECKPOINT_FREQUENCY=8
```

This modification causes the standalone network to close ledgers every 1 second and create a checkpoint once every 8 ledgers, 
deviating from the default timing of ledger closing after 5 seconds and creating a checkpoint once every 64 ledgers. Please note that 
this customization is only applicable when running a standalone network, as it requires changes to the captive-core configuration.

## Using a specific version of Stellar Core

> By default, the Docker Compose file is configured to use version 19 of Protocol and Stellar Core. You want the Core version to be at same level as the version horizon repo expects for ingestion. You can specify optional environment variables from the command shell for stating version overrides for either the docker-compose or start.sh invocations.

Reviewer 2: Not sure what you mean by "the version horizon repo expects"? How does one find the expected version?

Author: I will update it to whatever is specified in horizon.yml

Author: This part was already present in the documentation. I will check to see where to find a particular version horizon uses or remove this sentence altogether.

```bash
export PROTOCOL_VERSION="19"
export CORE_IMAGE="stellar/stellar-core:19.11.0-1323.7fb6d5e88.focal" 
export STELLAR_CORE_VERSION="19.11.0-1323.7fb6d5e88.focal"
```

Example:

Runs Stellar Protocol and Core version 19, for any mode of testnet, standalone, pubnet
```bash
PROTOCOL_VERSION=19 CORE_IMAGE=stellar/stellar-core:19.11.0-1323.7fb6d5e88.focal STELLAR_CORE_VERSION=19.11.0-1323.7fb6d5e88.focal ./start.sh [standalone|pubnet]
```

## <a name="logging"></a> **Logging**

> All logging infrastructure is in the `go/services/horizon/internal/log` package.  This package provides "level-based" logging:  Each logging statement has a severity, one of "Debug", "Info", "Warn", "Error" or "Panic".  The Horizon server has a configured level "filter", specified either using the `--log-level` command line flag or the `LOG_LEVEL` environment variable.  When a logging statement is executed, the statements declared severity is checked against the filter and will only be emitted if the severity of the statement is equal or higher severity than the filter.

Reviewer: not sure where this reference for log package comes, looking at the internal ingest code for examples of the logger used:

https://github.com/stellar/go/blob/master/services/horizon/internal/ingest/main.go#L72

it looks like horizon internal packages use an instance of the logger from the general `github.com/stellar/go/support/log` pkg in the go repo. I think can just change reference to that package here, and the rest of severity level and custom fields is still accurate.
